File: The Noob Bot/nManager.QPE/Control/ucXmlRichTextBox.vb
Imports System.Runtime.CompilerServices
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Xml.Linq

Namespace CustomXmlViewer
	Public Partial Class ucXmlRichTextBox
		Inherits RichTextBox
		#Region "Constructor"
		Public Sub New()
			InitializeComponent()

			Me.Font = New System.Drawing.Font("Consolas", 9.5F, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, CByte(0))

            AddHandler miCopyRtf.Click, AddressOf miCopyRtf_Click
            AddHandler miCopyText.Click, AddressOf miCopyText_Click
            AddHandler miSelectAll.Click, AddressOf miSelectAll_Click


        End Sub
		#End Region

		#Region "Properties"
		Private m_xml As String = ""
		Public Property Xml() As String
			Get
				Return m_xml
			End Get
			Set
				Me.Text = ""
				m_xml = value
				SetXml(m_xml)
			End Set
		End Property
		#End Region

		#Region "Private Methods"
		Private Sub SetXml(s As String)
			If [String].IsNullOrEmpty(s) Then
				Return
			End If

			Dim xdoc As XDocument = XDocument.Parse(s)

			Dim formattedText As String = xdoc.ToString().Trim()

			If [String].IsNullOrEmpty(formattedText) Then
				Return
			End If

			Dim machine As New XmlStateMachine()

			If s.StartsWith("<?") Then
				Dim xmlDeclaration As String = machine.GetXmlDeclaration(s)
				If xmlDeclaration <> [String].Empty Then
					formattedText = xmlDeclaration + Environment.NewLine & formattedText
				End If
			End If

			Dim location As Integer = 0
			Dim failCount As Integer = 0
			Dim tokenTryCount As Integer = 0
			Dim ttype As XmlTokenType = XmlTokenType.Unknown
			While location < formattedText.Length
				Dim token As String = machine.GetNextToken(formattedText, location, ttype)
				Dim color As Color = machine.GetTokenColor(ttype)
				Me.AppendText(token, color)
				location += token.Length
				tokenTryCount += 1

				' Check for ongoing failure
				If token.Length = 0 Then
					failCount += 1
				End If
				If failCount > 10 OrElse tokenTryCount > formattedText.Length Then
					Dim theRestOfIt As String = formattedText.Substring(location, formattedText.Length - location)
					'this.AppendText(Environment.NewLine + Environment.NewLine + theRestOfIt); // DEBUG
					Me.AppendText(theRestOfIt)
					Exit While
				End If
			End While
		End Sub
		#End Region

		#Region "Context Menu"
		Private Sub contextMenuStrip1_Opening(sender As Object, e As CancelEventArgs)
			If [String].IsNullOrEmpty(Me.SelectedText) Then
				miCopyText.Enabled = False
				miCopyRtf.Enabled = False
			Else
				miCopyText.Enabled = True
				miCopyRtf.Enabled = True
			End If
		End Sub

		Private Sub miCopyText_Click(sender As Object, e As EventArgs)
			Dim s As String = Me.SelectedText
			Try
				Dim doc As XDocument = XDocument.Parse(s)
				s = doc.ToString()
			Catch
			End Try
			Clipboard.SetText(s)
		End Sub

		Private Sub miCopyRtf_Click(sender As Object, e As EventArgs)
			Dim dto As New DataObject()
			dto.SetText(Me.SelectedRtf, TextDataFormat.Rtf)
			dto.SetText(Me.SelectedText, TextDataFormat.UnicodeText)
			Clipboard.Clear()
			Clipboard.SetDataObject(dto)
		End Sub

		Private Sub miSelectAll_Click(sender As Object, e As EventArgs)
			Me.SelectAll()
		End Sub
		#End Region
	End Class

#Region "Extension Methods"
    Module RichTextBoxExtensions

        <Extension()>
        Public Sub AppendText(box As RichTextBox, text As String, color As Color)
            box.SelectionStart = box.TextLength
            box.SelectionLength = 0

            box.SelectionColor = color
            box.AppendText(text)
            box.SelectionColor = box.ForeColor
        End Sub
    End Module
#End Region
End Namespace

'=======================================================
'Service provided by Telerik (www.telerik.com)
'Conversion powered by NRefactory.
'Twitter: @telerik
'Facebook: facebook.com/telerik
'=======================================================

File: The Noob Bot/nManager/Helpful/Forms/UserControls/ucXmlRichTextBox.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using System.Xml.Linq;

namespace nManager.Helpful.Forms.UserControls
{
    public sealed partial class UCXmlRichTextBox : RichTextBox
    {
        #region Constructor

        public UCXmlRichTextBox()
        {
            InitializeComponent();

            Font = new Font("Consolas", 9.5F, FontStyle.Regular, GraphicsUnit.Point, 0);

            miCopyRtf.Click += miCopyRtf_Click;
            miCopyText.Click += miCopyText_Click;
            miSelectAll.Click += miSelectAll_Click;
        }

        #endregion Constructor

        #region Properties

        private string _xml = "";

        public string Xml
        {
            get { return _xml; }
            set
            {
                Text = "";
                _xml = value;
                SetXml(_xml);
            }
        }

        #endregion Properties

        #region Private Methods

        private void SetXml(string s)
        {
            if (String.IsNullOrEmpty(s)) return;

            XDocument xdoc = XDocument.Parse(s);

            string formattedText = xdoc.ToString().Trim();

            if (String.IsNullOrEmpty(formattedText)) return;

            var machine = new XmlStateMachine();

            if (s.StartsWith("<?"))
            {
                string xmlDeclaration = machine.GetXmlDeclaration(s);
                if (xmlDeclaration != String.Empty) formattedText = xmlDeclaration + Environment.NewLine + formattedText;
            }

            int location = 0;
            int failCount = 0;
            int tokenTryCount = 0;
            while (location < formattedText.Length)
            {
                XmlTokenType ttype;
                string token = machine.GetNextToken(formattedText, location, out ttype);
                Color color = machine.GetTokenColor(ttype);
                AppendText(this, token, color);

                location += token.Length;
                tokenTryCount++;

                // Check for ongoing failure
                if (token.Length == 0) failCount++;
                if (failCount > 10 || tokenTryCount > formattedText.Length)
                {
                    string theRestOfIt = formattedText.Substring(location, formattedText.Length - location);
                    //this.AppendText(Environment.NewLine + Environment.NewLine + theRestOfIt); // DEBUG
                    AppendText(theRestOfIt);
                    break;
                }
            }
        }

        #endregion Private Methods

        public static void AppendText(RichTextBox box, string text, Color color)
        {
            box.SelectionStart = box.TextLength;
            box.SelectionLength = 0;

            box.SelectionColor = color;
            box.AppendText(text);
            box.SelectionColor = box.ForeColor;
        }

        #region Context Menu

        private void contextMenuStrip1_Opening(object sender, CancelEventArgs e)
        {
            if (String.IsNullOrEmpty(SelectedText))
            {
                miCopyText.Enabled = false;
                miCopyRtf.Enabled = false;
            }
            else
            {
                miCopyText.Enabled = true;
                miCopyRtf.Enabled = true;
            }
        }

        private void miCopyText_Click(object sender, EventArgs e)
        {
            string s = SelectedText;
            try
            {
                XDocument doc = XDocument.Parse(s);
                s = doc.ToString();
            }
            catch
            {
            }
            Clipboard.SetText(s);
        }

        private void miCopyRtf_Click(object sender, EventArgs e)
        {
            var dto = new DataObject();
            dto.SetText(SelectedRtf, TextDataFormat.Rtf);
            dto.SetText(SelectedText, TextDataFormat.UnicodeText);
            Clipboard.Clear();
            Clipboard.SetDataObject(dto);
        }

        private void miSelectAll_Click(object sender, EventArgs e)
        {
            SelectAll();
        }

        #endregion Context Menu
    }

    #region Extension Methods

    internal static class RichTextBoxExtensions
    {
        public static void AppendText(this RichTextBox box, string text, Color color)
        {
            box.SelectionStart = box.TextLength;
            box.SelectionLength = 0;

            box.SelectionColor = color;
            box.AppendText(text);
            box.SelectionColor = box.ForeColor;
        }
    }

    #endregion Extension Methods
}